### ISSUE TITLE
Write hashes of nar chunks into sqlite

### ISSUE THREAD

**Issue by volth** [NixOS/nix#3260]
It would be nice is `Nix` will calculate and keep in its `sqlite` database not only `narHash` - `sha256 `of .nar file of each derivation, but also `sha1` of each 4MiB chunk of the same nar file.

Why? - to create something like https://github.com/edolstra/nix-serve but with BitTorrent instead of HTTP.

Of course, hashes of chunks could be calculated and maintained by external software, but it has disadvantages:
1. The external tool will have to re-read whole nix store (on contrary, Nix can calculate both hashes - `narHash` and chunk hashes simultaneously, without extra reading)
2. The external tool will introduce delay. Freshly build derivation is immediately available via HTTP but it will not be so for BitTorrent.

**Comment by volth** [comment]
It is only about improving nixstore-to-nixstore transfers (next gen of `nix-serve` and `nix-copy-closure` with P2P and resuming), not about distributing compressed `.nar.xz` from Hydra.

**Comment by edolstra** [comment]
I don't see how this would work in practice. The problem is that the chunk hashes don't allow random access - given a chunk hash, I can't see an efficient way to get the corresponding substring of the NAR serialization of the path.

It also seems like a bit of a layering violation to store BitTorrent hashes directly in the Nix database.

**Comment by volth** [comment]
> given a chunk hash, I can't see an efficient way to get the corresponding substring of the NAR serialization of the path.

Lookup for a chunk hash in sqlite will give the derivation and number of the chunk (= offset in .nar file).
Making a random chunk will either require directory traversal, although without reading skipped files, or the same table might contain some hints to speed up the skipping.

**Comment by volth** [comment]
> It also seems like a bit of a layering violation to store BitTorrent hashes directly in the Nix database.

magnet hash (approx: sha1 of concatenated string of sha1 of chunks) duplicates `narHash` (sha256 of whole .nar file), so classic `narHash` might be gone as less useful.

**Comment by tomberek** [comment]
I've distributed some large mostly-similar files using http://0pointer.net/blog/casync-a-tool-for-distributing-file-system-images.html (go implementation: https://github.com/folbricht/desync). It uses a rolling hash scheme that can find many more opportunities for sharing than a straight chunked hash. The individual chunks are still addressed via sha256 and stored compressed. Something like that can work well with torrents (my interest was to eventually get it working via WebTorrents).

It introduces a new serialization format like nar, the format subsumes tar, has random-access and is designed to be reproducible. Made me think that just as we have alternate hashes supported by SRI, additional archive formats may be supported if it is determined that the additional features are worth it. (or find good ideas and re-use them)

I've not tried, but also thought about using a nix store to serve the content addressed chunks instead of casync's own store format.

Other random things:
- built a representation of the format in Kaitai which can generate parsing code in various languages
- can do chunking and hashing in parallel, this is a big time saver for large files
- improve transfer of large files and reduce the cost of failed transfers. I've done this via a "fetchUrlChunked" function in some projects that uses "Range: bytes=0-1073741823" headers in curlOpts to improve initial import of multi-GiB files to a nix store. (side question: would this be valuable to add anywhere?)

**Comment by volth** [comment]
@tomberek yes, there is also a possibility to check individual files in `/nix/store/.links` to avoid transfer what is already on the target.
Those optimizations are nice, but slightly offtopic. My first goal is to get rid of `http` (and probably `ssh`) with their connection aborts after an arbitrary chosen timeouts (I hit them too often - not only when am using bad last mile connection, but also in datacenters, because cache server's load has spikes so its network is saturated), exact endpoint addresses chosen beforehand (which then result in ugly hacks like assigning priorities to each of binary cache server), etc

**Comment by wamserma** [comment]
@voith what you are looking after is more like having an IPFS-based cache. People have been working on that but then again dropped it, because of IPFS' abysmal performance.
On the other hand `casync` mentioned by @tomberek uses a serialisation format that is much more complex that `nar` (but similar in spirit). If you take out the serialisation, what is left is quite similar to `zsync`. The  zsync-file can simply accompany the existing narinfo and range queries should be supported by any reasonable CDN nowadays.
The main dealbraker for `zsync` is that it only supports `gzip` compression, and I don't know if that can be updated to supporting `xz`or `zstd`.
This discussion should probably taken to discourse, but loading the sqlite with chunk hashes is probably not the way to go. I'd rather see some form of plugin-support to nix, where you can hook in stuff (just a bit like apt-transport).

**Comment by volth** [comment]
I'd say https://github.com/uber/kraken is much closer to what I want than IPFS

**Comment by voith** [comment]
> @voith what you are looking after is more like having an IPFS-based cache.

@wamserma I'm guessing that you meant to ping @ volth and not me(@voith)
